<!doctype html>
<html lang="en">
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, minimum-scale=1" />
<meta name="generator" content="pdoc 0.7.1" />
<title>miniDna.sequence API documentation</title>
<meta name="description" content="" />
<link href='https://cdnjs.cloudflare.com/ajax/libs/normalize/8.0.0/normalize.min.css' rel='stylesheet'>
<link href='https://cdnjs.cloudflare.com/ajax/libs/10up-sanitize.css/8.0.0/sanitize.min.css' rel='stylesheet'>
<link href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/github.min.css" rel="stylesheet">
<style>.flex{display:flex !important}body{line-height:1.5em}#content{padding:20px}#sidebar{padding:30px;overflow:hidden}.http-server-breadcrumbs{font-size:130%;margin:0 0 15px 0}#footer{font-size:.75em;padding:5px 30px;border-top:1px solid #ddd;text-align:right}#footer p{margin:0 0 0 1em;display:inline-block}#footer p:last-child{margin-right:30px}h1,h2,h3,h4,h5{font-weight:300}h1{font-size:2.5em;line-height:1.1em}h2{font-size:1.75em;margin:1em 0 .50em 0}h3{font-size:1.4em;margin:25px 0 10px 0}h4{margin:0;font-size:105%}a{color:#058;text-decoration:none;transition:color .3s ease-in-out}a:hover{color:#e82}.title code{font-weight:bold}h2[id^="header-"]{margin-top:2em}.ident{color:#900}pre code{background:#f8f8f8;font-size:.8em;line-height:1.4em}code{background:#f2f2f1;padding:1px 4px;overflow-wrap:break-word}h1 code{background:transparent}pre{background:#f8f8f8;border:0;border-top:1px solid #ccc;border-bottom:1px solid #ccc;margin:1em 0;padding:1ex}#http-server-module-list{display:flex;flex-flow:column}#http-server-module-list div{display:flex}#http-server-module-list dt{min-width:10%}#http-server-module-list p{margin-top:0}.toc ul,#index{list-style-type:none;margin:0;padding:0}#index code{background:transparent}#index h3{border-bottom:1px solid #ddd}#index ul{padding:0}#index h4{font-weight:bold}#index h4 + ul{margin-bottom:.6em}@media (min-width:200ex){#index .two-column{column-count:2}}@media (min-width:300ex){#index .two-column{column-count:3}}dl{margin-bottom:2em}dl dl:last-child{margin-bottom:4em}dd{margin:0 0 1em 3em}#header-classes + dl > dd{margin-bottom:3em}dd dd{margin-left:2em}dd p{margin:10px 0}.name{background:#eee;font-weight:bold;font-size:.85em;padding:5px 10px;display:inline-block;min-width:40%}.name:hover{background:#e0e0e0}.name > span:first-child{white-space:nowrap}.name.class > span:nth-child(2){margin-left:.4em}.inherited{color:#999;border-left:5px solid #eee;padding-left:1em}.inheritance em{font-style:normal;font-weight:bold}.desc h2{font-weight:400;font-size:1.25em}.desc h3{font-size:1em}.desc dt code{background:inherit}.source summary,.git-link-div{color:#666;text-align:right;font-weight:400;font-size:.8em;text-transform:uppercase}.source summary > *{white-space:nowrap;cursor:pointer}.git-link{color:inherit;margin-left:1em}.source pre{max-height:500px;overflow:auto;margin:0}.source pre code{font-size:12px;overflow:visible}.hlist{list-style:none}.hlist li{display:inline}.hlist li:after{content:',\2002'}.hlist li:last-child:after{content:none}.hlist .hlist{display:inline;padding-left:1em}img{max-width:100%}.admonition{padding:.1em .5em;margin-bottom:1em}.admonition-title{font-weight:bold}.admonition.note,.admonition.info,.admonition.important{background:#aef}.admonition.todo,.admonition.versionadded,.admonition.tip,.admonition.hint{background:#dfd}.admonition.warning,.admonition.versionchanged,.admonition.deprecated{background:#fd4}.admonition.error,.admonition.danger,.admonition.caution{background:lightpink}</style>
<style media="screen and (min-width: 700px)">@media screen and (min-width:700px){#sidebar{width:30%}#content{width:70%;max-width:100ch;padding:3em 4em;border-left:1px solid #ddd}pre code{font-size:1em}.item .name{font-size:1em}main{display:flex;flex-direction:row-reverse;justify-content:flex-end}.toc ul ul,#index ul{padding-left:1.5em}.toc > ul > li{margin-top:.5em}}</style>
<style media="print">@media print{#sidebar h1{page-break-before:always}.source{display:none}}@media print{*{background:transparent !important;color:#000 !important;box-shadow:none !important;text-shadow:none !important}a[href]:after{content:" (" attr(href) ")";font-size:90%}a[href][title]:after{content:none}abbr[title]:after{content:" (" attr(title) ")"}.ir a:after,a[href^="javascript:"]:after,a[href^="#"]:after{content:""}pre,blockquote{border:1px solid #999;page-break-inside:avoid}thead{display:table-header-group}tr,img{page-break-inside:avoid}img{max-width:100% !important}@page{margin:0.5cm}p,h2,h3{orphans:3;widows:3}h1,h2,h3,h4,h5,h6{page-break-after:avoid}}</style>
</head>
<body>
<main>
<article id="content">
<header>
<h1 class="title">Module <code>miniDna.sequence</code></h1>
</header>
<section id="section-intro">
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python"># -*- coding: utf-8 -*-

import math

# all nucleotides
NUCLEOTIDES = &#39;ATGCUatgcu&#39;

# codon to amino acid table
AMINOCODE =  {
  &#39;ATG&#39;: &#39;M&#39;, &#39;GCG&#39;: &#39;A&#39;, &#39;TCA&#39;: &#39;S&#39;,
  &#39;GAA&#39;: &#39;E&#39;, &#39;GGG&#39;: &#39;G&#39;, &#39;GGT&#39;: &#39;G&#39;,
  &#39;AAA&#39;: &#39;K&#39;, &#39;GAG&#39;: &#39;E&#39;, &#39;AAT&#39;: &#39;N&#39;,
  &#39;CTA&#39;: &#39;L&#39;, &#39;CAT&#39;: &#39;H&#39;, &#39;TCG&#39;: &#39;S&#39;,
  &#39;TAG&#39;: &#39;STOP&#39;, &#39;GTG&#39;: &#39;V&#39;, &#39;TAT&#39;: &#39;Y&#39;,
  &#39;CCT&#39;: &#39;P&#39;, &#39;ACT&#39;: &#39;T&#39;, &#39;TCC&#39;: &#39;S&#39;,
  &#39;CAG&#39;: &#39;Q&#39;, &#39;CCA&#39;: &#39;P&#39;, &#39;TAA&#39;: &#39;STOP&#39;,
  &#39;AGA&#39;: &#39;R&#39;, &#39;ACG&#39;: &#39;T&#39;, &#39;CAA&#39;: &#39;Q&#39;,
  &#39;TGT&#39;: &#39;C&#39;, &#39;GCT&#39;: &#39;A&#39;, &#39;TTC&#39;: &#39;F&#39;,
  &#39;AGT&#39;: &#39;S&#39;, &#39;ATA&#39;: &#39;I&#39;, &#39;TTA&#39;: &#39;L&#39;,
  &#39;CCG&#39;: &#39;P&#39;, &#39;ATC&#39;: &#39;I&#39;, &#39;TTT&#39;: &#39;F&#39;,
  &#39;CGT&#39;: &#39;R&#39;, &#39;TGA&#39;: &#39;STOP&#39;, &#39;GTA&#39;: &#39;V&#39;,
  &#39;TCT&#39;: &#39;S&#39;, &#39;CAC&#39;: &#39;H&#39;, &#39;GTT&#39;: &#39;V&#39;,
  &#39;GAT&#39;: &#39;D&#39;, &#39;CGA&#39;: &#39;R&#39;, &#39;GGA&#39;: &#39;G&#39;,
  &#39;GTC&#39;: &#39;V&#39;, &#39;GGC&#39;: &#39;G&#39;, &#39;TGC&#39;: &#39;C&#39;,
  &#39;CTG&#39;: &#39;L&#39;, &#39;CTC&#39;: &#39;L&#39;, &#39;CGC&#39;: &#39;R&#39;,
  &#39;CGG&#39;: &#39;R&#39;, &#39;AAC&#39;: &#39;N&#39;, &#39;GCC&#39;: &#39;A&#39;,
  &#39;ATT&#39;: &#39;I&#39;, &#39;AGG&#39;: &#39;R&#39;, &#39;GAC&#39;: &#39;D&#39;,
  &#39;ACC&#39;: &#39;T&#39;, &#39;AGC&#39;: &#39;S&#39;, &#39;TAC&#39;: &#39;Y&#39;,
  &#39;ACA&#39;: &#39;T&#39;, &#39;AAG&#39;: &#39;K&#39;, &#39;GCA&#39;: &#39;A&#39;,
  &#39;TTG&#39;: &#39;L&#39;, &#39;CCC&#39;: &#39;P&#39;, &#39;CTT&#39;: &#39;L&#39;,
  &#39;TGG&#39;: &#39;W&#39;
}
&#34;&#34;&#34;Table to convert Codon to Amino Acid&#34;&#34;&#34;

AMINO = {
  &#39;A&#39;: &#39;Alanine&#39;,
  &#39;F&#39;: &#39;Phenyl-alanine&#39;,
  &#39;L&#39;: &#39;Leucine&#39;,
  &#39;P&#39;: &#39;Proline&#39;,
  &#39;H&#39;: &#39;Histidine&#39;,
  &#39;Q&#39;: &#39;Glutamine&#39;,
  &#39;R&#39;: &#39;Arginine&#39;,
  &#39;I&#39;: &#39;Isoleucine&#39;,
  &#39;T&#39;: &#39;Threonine&#39;,
  &#39;N&#39;: &#39;Asparagine&#39;,
  &#39;K&#39;: &#39;Lysine&#39;,
  &#39;S&#39;: &#39;Serine&#39;,
  &#39;V&#39;: &#39;Valine&#39;,
  &#39;D&#39;: &#39;Aspartic acid&#39;,
  &#39;E&#39;: &#39;Glutamic acid&#39;,
  &#39;G&#39;: &#39;Glycine&#39;,
  &#39;M&#39;: &#39;Methionine&#39;,
  &#39;W&#39;: &#39;Trytophane&#39;,
  &#39;Y&#39;: &#39;Tyrosine&#39;,
  &#39;C&#39;: &#39;Cystein&#39;
}
&#34;&#34;&#34;Table of all Amino Acid&#34;&#34;&#34;


def isDna(seq: str, msg: bool = False) -&gt; bool: 
  &#34;&#34;&#34;Test if a string is a DNA sequence.

    **Keyword arguments:**  
    seq -- the string to test  
    msg -- display messages or not
  &#34;&#34;&#34;
  isRna = False
  for n in seq:
    if n not in NUCLEOTIDES:
      print(&#34;Invalid DNA sequence&#34;)
      return False
    elif n == &#39;U&#39; or n == &#39;u&#39;:
      isRna = True

  if isRna and msg:
    print(seq + &#34;\nis a RNA sequence&#34;)
  elif msg:
    print(seq + &#34;\nis a DNA sequence&#34;)
  return True


def percentIdentical(seqA: str, seqB: str) -&gt; float:
  &#34;&#34;&#34;Return the percentage of identity between two DNA sequences.
    
    **Keyword arguments:**  
    seqA -- the first sequence
    seqB -- the second sequence
  &#34;&#34;&#34;

  if len(seqA) != len(seqB):
    raise ValueError(&#34;sequences have unequal length&#34;)
  dist = sum(ch1 != ch2 for ch1, ch2 in zip(seqA, seqB))
  return math.ceil(100 * (1 - dist/len(seqA)))


def countIdentical(seqA: str, seqB: str) -&gt; int:
  &#34;&#34;&#34;Return the number of equal nucleotides between two sequences.

    **Keyword arguments:**  
    seqA -- the first sequence  
    seqB -- the second sequence
  &#34;&#34;&#34;

  if len(seqA) != len(seqB):
    raise ValueError(&#34;sequences have unequal length&#34;)
  count = sum(ch1 == ch2 for ch1, ch2 in zip(seqA, seqB))
  return count


def simpleAlign(seqA: str, seqB: str) -&gt; tuple:
  &#34;&#34;&#34;Find the best way to superimpose two sequences
    without changing them.
    Empty nucleotides used to match as best as possible
    the sequences are symbolized with a &#34;-&#34; character.
    
    **Keyword arguments:**  
    seqA -- the first sequence  
    seqB -- the second sequence
  &#34;&#34;&#34;

  bestShift = 0
  shift = 0
  countId = countIdentical(seqA, seqB)

  while(shift &lt; len(seqA)):
    shift += 1
    a = seqA[shift:len(seqA)]
    b = seqB[0:len(seqB) - shift]
    c = countIdentical(a, b)
    if c &gt; countId:
      countId = c
      bestShift = shift

  
  finalA = seqA + &#34;&#34;.join(&#34;-&#34; for j in range(bestShift))
  finalB = &#34;&#34;.join(&#34;-&#34; for i in range(bestShift)) + seqB

  return (finalA, finalB)
  

def _subs(a: str, b: str, ide: int = 3, sub: int = -1) -&gt; int:
  return (a == b) * ide + (a != b) * sub


def globalAlign(seqA: str, seqB: str, ide: int = 3, sub: int = -1, ind: int = -3) -&gt; tuple:
  &#34;&#34;&#34;Find the best global alignement between two sequences.
    Return the two aligned sequences as a tuple.

    **Keyword arguments:**  
    seqA -- first sequence  
    seqB -- second sequence  
    ide -- identity score  
    sub -- substitution score  
    ind -- indel score  
  &#34;&#34;&#34;

  F = [[0 for x in range(len(seqA) + 1)] for y in range(len(seqB) + 1)]
  
  for i in range(len(seqA) + 1):
    F[0][i] = ind * i
  for i in range(len(seqB) + 1):
    F[i][0] = ind * i

  for i in range(1, len(seqB) + 1):
    for j in range(1, len(seqA) + 1):
      s1 = F[i-1][j-1] + _subs(seqA[j-1], seqB[i-1], ide, sub)
      s2 = F[i-1][j] + ind
      s3 = F[i][j-1] + ind
      F[i][j] = max(s1, s2, s3)

  alignA = &#34;&#34;
  alignB = &#34;&#34;

  i = len(seqB)
  j = len(seqA)

  while(i &gt; 0 and j &gt; 0):
    score = F[i][j]
    scoreDiag = F[i-1][j-1]
    scoreUp = F[i-1][j]
    scoreLeft = F[i][j-1]
    if score == scoreDiag + _subs(seqA[j-1], seqB[i-1], ide, sub):
      alignA = seqA[j-1] + alignA
      alignB = seqB[i-1] + alignB
      i -= 1
      j -= 1
    elif score == scoreLeft + ind:
      alignA = seqA[j-1] + alignA
      alignB = &#34;-&#34; + alignB
      j -= 1
    elif score == scoreUp + ind:
      alignA = &#34;-&#34; + alignA
      alignB = seqB[i-1] + alignB
      i -= 1

  while(j &gt; 0):
    alignA = seqA[j-1] + alignA
    alignB = &#34;-&#34; + alignB
    j -= 1
  while(i &gt; 0):
    alignA = &#34;-&#34; + alignA
    alignB = seqB[i-1] + alignB
    i -= 1

  print(alignA)
  print(alignB)
  return (alignA, alignB)


def freqList(seqList: list, prob: bool = True) -&gt; dict:
  &#34;&#34;&#34;Return the frequency of each nucleotide in each
    position.

    **Keyword arguments:**  
    seqList -- DNA sequences list  
    prob -- False: return a dictionnary of probability (between 0 and 1)
            True: return a dictionnary of freqency (int)
            Default is True
  &#34;&#34;&#34;

  n = len(seqList[0])
  nl = len(seqList)
  freqs = {
    &#39;A&#39;: [0]*n,
    &#39;T&#39;: [0]*n,
    &#39;G&#39;: [0]*n,
    &#39;C&#39;: [0]*n
  }
  for seq in seqList:
    for i, nuc in enumerate(seq):
        freqs[nuc][i] += 1
  if prob:
    for key in freqs:
      for i in range(len(freqs[key])):
        freqs[key][i] = freqs[key][i]/nl
  return freqs


def freqAt(freqDict: dict, nuc: str, n: int) -&gt; float:
  &#34;&#34;&#34;Read a frequency dictionnary returned 
    by the function freqList.

    **Keyword arguments:**  
    freqDic -- frequence dictionnary  
    nuc -- nucleotide A,T,G or C  
    n -- position
  &#34;&#34;&#34;

  return freqDict[nuc][n]


def translate(seq: str) -&gt; str:
  &#34;&#34;&#34;find the protein coded in a DNA sequence
    
    **Keyword arguments:**  
    seq -- DNA sequence to translate  
  &#34;&#34;&#34;

  start = 0
  protein = &#39;&#39;
  while start+2 &lt; len(seq):
    codon = seq[start:start+3]
    protein += AMINOCODE[codon]
    start += 3
  return protein


def dotPlot(seqA: str, seqB: str) -&gt; list:
  &#34;&#34;&#34;Compare two DNA sequences using DotPlot method.

    **Keyword arguments:**  
    seqA -- first sequence  
    seqB -- second sequence  
  &#34;&#34;&#34;

  la = len(seqA)
  lb = len(seqB)
  M = [[&#39; &#39; for n in range(la)] for m in range(lb)]
  for i in range(lb):
    for j in range(la):
      if seqB[i] == seqA[j]:
        M[i][j] = &#39;x&#39;
  return M


def filterDotPlot(seqA: str, seqB: str, k: int) -&gt; list:
  &#34;&#34;&#34;Compare two DNA sequences using DotPlot method 
    keep only K long equals sequences

    **Keyword arguments:**  
    seqA -- first sequence  
    seqB -- second sequence  
    k -- minimum length of equal portions  
  &#34;&#34;&#34;

  if isDna(seqA) and isDna(seqB):
    la = len(seqA)
    lb = len(seqB)
    M = [[&#39; &#39; for n in range(la)] for m in range(lb)]
    for i in range(lb + 1 - k):
      for j in range(la + 1 - k):
        a = seqA[j:j+k]
        b = seqB[i:i+k]
        if a == b:
          for d in range(k):
            M[i+d][j+d] = &#39;x&#39;
    return M


def compare(seqA: str, seqB: str, verbose: bool = False) -&gt; None:
  &#34;&#34;&#34;Compare two sequences.  
    Display the result in the console

    **Keyword arguments:**  
    seqA -- first sequence (DNA or Protein)  
    seqB -- first sequence (DNA or Protein)   
    verbose -- display or not sequences in the console
  &#34;&#34;&#34;

  la = len(seqA)
  sim = &#39;&#39;
  match = 0
  diff = 0

  for n in range(la):
    if seqA[n] == seqB[n]:
      sim += &#39;|&#39;
      match += 1
    else:
      sim += &#39; &#39;
      diff += 1

  if verbose:
    print(&#39;======|=&#39; + &#39;=&#39; * len(seqA))
    print(&#39; seqA | &#39; + seqA)
    print(&#39; ---- | &#39; + sim)
    print(&#39; seqB | &#39; + seqB)
    print(&#39;======|=&#39; + &#39;=&#39; * len(seqA))

  print(&#39;======|&gt; &#39; + str(diff) + &#39; differences&#39;)
  print(&#39;======|&gt; &#39; + str(match) + &#39; matches&#39;)
  print(&#39;======|&gt; &#39; + str(percentIdentical(seqA, seqB)) + &#39;% identity&#39;)


def display(a: str, b: str, m: list) -&gt; None:
  &#34;&#34;&#34;Display dotplot matrix properly
    
    **Keyword arguments:**  
    a -- first sequence compared  
    b -- second sequence compared  
    m -- dotplot matrix  
  &#34;&#34;&#34;
  header = &#39;  &#39;

  for c in a:
    header += (c + &#39; &#39;)

  print(header)
  print(&#39;==&#39; * len(a) + &#39;===&#39;)

  for i in range(len(m)):
    s = b[i] + &#39; &#39;
    for j in range(len(m[0])):
      s += (m[i][j] + &#39; &#39;)
    print(s)</code></pre>
</details>
</section>
<section>
</section>
<section>
<h2 class="section-title" id="header-variables">Global variables</h2>
<dl>
<dt id="miniDna.sequence.AMINO"><code class="name">var <span class="ident">AMINO</span></code></dt>
<dd>
<section class="desc"><p>Table of all Amino Acid</p></section>
</dd>
<dt id="miniDna.sequence.AMINOCODE"><code class="name">var <span class="ident">AMINOCODE</span></code></dt>
<dd>
<section class="desc"><p>Table to convert Codon to Amino Acid</p></section>
</dd>
</dl>
</section>
<section>
<h2 class="section-title" id="header-functions">Functions</h2>
<dl>
<dt id="miniDna.sequence.compare"><code class="name flex">
<span>def <span class="ident">compare</span></span>(<span>seqA, seqB, verbose=False)</span>
</code></dt>
<dd>
<section class="desc"><p>Compare two sequences.<br>
Display the result in the console</p>
<p><strong>Keyword arguments:</strong><br>
seqA &ndash; first sequence (DNA or Protein)<br>
seqB &ndash; first sequence (DNA or Protein) <br>
verbose &ndash; display or not sequences in the console</p></section>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def compare(seqA: str, seqB: str, verbose: bool = False) -&gt; None:
  &#34;&#34;&#34;Compare two sequences.  
    Display the result in the console

    **Keyword arguments:**  
    seqA -- first sequence (DNA or Protein)  
    seqB -- first sequence (DNA or Protein)   
    verbose -- display or not sequences in the console
  &#34;&#34;&#34;

  la = len(seqA)
  sim = &#39;&#39;
  match = 0
  diff = 0

  for n in range(la):
    if seqA[n] == seqB[n]:
      sim += &#39;|&#39;
      match += 1
    else:
      sim += &#39; &#39;
      diff += 1

  if verbose:
    print(&#39;======|=&#39; + &#39;=&#39; * len(seqA))
    print(&#39; seqA | &#39; + seqA)
    print(&#39; ---- | &#39; + sim)
    print(&#39; seqB | &#39; + seqB)
    print(&#39;======|=&#39; + &#39;=&#39; * len(seqA))

  print(&#39;======|&gt; &#39; + str(diff) + &#39; differences&#39;)
  print(&#39;======|&gt; &#39; + str(match) + &#39; matches&#39;)
  print(&#39;======|&gt; &#39; + str(percentIdentical(seqA, seqB)) + &#39;% identity&#39;)</code></pre>
</details>
</dd>
<dt id="miniDna.sequence.countIdentical"><code class="name flex">
<span>def <span class="ident">countIdentical</span></span>(<span>seqA, seqB)</span>
</code></dt>
<dd>
<section class="desc"><p>Return the number of equal nucleotides between two sequences.</p>
<p><strong>Keyword arguments:</strong><br>
seqA &ndash; the first sequence<br>
seqB &ndash; the second sequence</p></section>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def countIdentical(seqA: str, seqB: str) -&gt; int:
  &#34;&#34;&#34;Return the number of equal nucleotides between two sequences.

    **Keyword arguments:**  
    seqA -- the first sequence  
    seqB -- the second sequence
  &#34;&#34;&#34;

  if len(seqA) != len(seqB):
    raise ValueError(&#34;sequences have unequal length&#34;)
  count = sum(ch1 == ch2 for ch1, ch2 in zip(seqA, seqB))
  return count</code></pre>
</details>
</dd>
<dt id="miniDna.sequence.display"><code class="name flex">
<span>def <span class="ident">display</span></span>(<span>a, b, m)</span>
</code></dt>
<dd>
<section class="desc"><p>Display dotplot matrix properly</p>
<p><strong>Keyword arguments:</strong><br>
a &ndash; first sequence compared<br>
b &ndash; second sequence compared<br>
m &ndash; dotplot matrix</p></section>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def display(a: str, b: str, m: list) -&gt; None:
  &#34;&#34;&#34;Display dotplot matrix properly
    
    **Keyword arguments:**  
    a -- first sequence compared  
    b -- second sequence compared  
    m -- dotplot matrix  
  &#34;&#34;&#34;
  header = &#39;  &#39;

  for c in a:
    header += (c + &#39; &#39;)

  print(header)
  print(&#39;==&#39; * len(a) + &#39;===&#39;)

  for i in range(len(m)):
    s = b[i] + &#39; &#39;
    for j in range(len(m[0])):
      s += (m[i][j] + &#39; &#39;)
    print(s)</code></pre>
</details>
</dd>
<dt id="miniDna.sequence.dotPlot"><code class="name flex">
<span>def <span class="ident">dotPlot</span></span>(<span>seqA, seqB)</span>
</code></dt>
<dd>
<section class="desc"><p>Compare two DNA sequences using DotPlot method.</p>
<p><strong>Keyword arguments:</strong><br>
seqA &ndash; first sequence<br>
seqB &ndash; second sequence</p></section>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def dotPlot(seqA: str, seqB: str) -&gt; list:
  &#34;&#34;&#34;Compare two DNA sequences using DotPlot method.

    **Keyword arguments:**  
    seqA -- first sequence  
    seqB -- second sequence  
  &#34;&#34;&#34;

  la = len(seqA)
  lb = len(seqB)
  M = [[&#39; &#39; for n in range(la)] for m in range(lb)]
  for i in range(lb):
    for j in range(la):
      if seqB[i] == seqA[j]:
        M[i][j] = &#39;x&#39;
  return M</code></pre>
</details>
</dd>
<dt id="miniDna.sequence.filterDotPlot"><code class="name flex">
<span>def <span class="ident">filterDotPlot</span></span>(<span>seqA, seqB, k)</span>
</code></dt>
<dd>
<section class="desc"><p>Compare two DNA sequences using DotPlot method
keep only K long equals sequences</p>
<p><strong>Keyword arguments:</strong><br>
seqA &ndash; first sequence<br>
seqB &ndash; second sequence<br>
k &ndash; minimum length of equal portions</p></section>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def filterDotPlot(seqA: str, seqB: str, k: int) -&gt; list:
  &#34;&#34;&#34;Compare two DNA sequences using DotPlot method 
    keep only K long equals sequences

    **Keyword arguments:**  
    seqA -- first sequence  
    seqB -- second sequence  
    k -- minimum length of equal portions  
  &#34;&#34;&#34;

  if isDna(seqA) and isDna(seqB):
    la = len(seqA)
    lb = len(seqB)
    M = [[&#39; &#39; for n in range(la)] for m in range(lb)]
    for i in range(lb + 1 - k):
      for j in range(la + 1 - k):
        a = seqA[j:j+k]
        b = seqB[i:i+k]
        if a == b:
          for d in range(k):
            M[i+d][j+d] = &#39;x&#39;
    return M</code></pre>
</details>
</dd>
<dt id="miniDna.sequence.freqAt"><code class="name flex">
<span>def <span class="ident">freqAt</span></span>(<span>freqDict, nuc, n)</span>
</code></dt>
<dd>
<section class="desc"><p>Read a frequency dictionnary returned
by the function freqList.</p>
<p><strong>Keyword arguments:</strong><br>
freqDic &ndash; frequence dictionnary<br>
nuc &ndash; nucleotide A,T,G or C<br>
n &ndash; position</p></section>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def freqAt(freqDict: dict, nuc: str, n: int) -&gt; float:
  &#34;&#34;&#34;Read a frequency dictionnary returned 
    by the function freqList.

    **Keyword arguments:**  
    freqDic -- frequence dictionnary  
    nuc -- nucleotide A,T,G or C  
    n -- position
  &#34;&#34;&#34;

  return freqDict[nuc][n]</code></pre>
</details>
</dd>
<dt id="miniDna.sequence.freqList"><code class="name flex">
<span>def <span class="ident">freqList</span></span>(<span>seqList, prob=True)</span>
</code></dt>
<dd>
<section class="desc"><p>Return the frequency of each nucleotide in each
position.</p>
<p><strong>Keyword arguments:</strong><br>
seqList &ndash; DNA sequences list<br>
prob &ndash; False: return a dictionnary of probability (between 0 and 1)
True: return a dictionnary of freqency (int)
Default is True</p></section>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def freqList(seqList: list, prob: bool = True) -&gt; dict:
  &#34;&#34;&#34;Return the frequency of each nucleotide in each
    position.

    **Keyword arguments:**  
    seqList -- DNA sequences list  
    prob -- False: return a dictionnary of probability (between 0 and 1)
            True: return a dictionnary of freqency (int)
            Default is True
  &#34;&#34;&#34;

  n = len(seqList[0])
  nl = len(seqList)
  freqs = {
    &#39;A&#39;: [0]*n,
    &#39;T&#39;: [0]*n,
    &#39;G&#39;: [0]*n,
    &#39;C&#39;: [0]*n
  }
  for seq in seqList:
    for i, nuc in enumerate(seq):
        freqs[nuc][i] += 1
  if prob:
    for key in freqs:
      for i in range(len(freqs[key])):
        freqs[key][i] = freqs[key][i]/nl
  return freqs</code></pre>
</details>
</dd>
<dt id="miniDna.sequence.globalAlign"><code class="name flex">
<span>def <span class="ident">globalAlign</span></span>(<span>seqA, seqB, ide=3, sub=-1, ind=-3)</span>
</code></dt>
<dd>
<section class="desc"><p>Find the best global alignement between two sequences.
Return the two aligned sequences as a tuple.</p>
<p><strong>Keyword arguments:</strong><br>
seqA &ndash; first sequence<br>
seqB &ndash; second sequence<br>
ide &ndash; identity score<br>
sub &ndash; substitution score<br>
ind &ndash; indel score</p></section>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def globalAlign(seqA: str, seqB: str, ide: int = 3, sub: int = -1, ind: int = -3) -&gt; tuple:
  &#34;&#34;&#34;Find the best global alignement between two sequences.
    Return the two aligned sequences as a tuple.

    **Keyword arguments:**  
    seqA -- first sequence  
    seqB -- second sequence  
    ide -- identity score  
    sub -- substitution score  
    ind -- indel score  
  &#34;&#34;&#34;

  F = [[0 for x in range(len(seqA) + 1)] for y in range(len(seqB) + 1)]
  
  for i in range(len(seqA) + 1):
    F[0][i] = ind * i
  for i in range(len(seqB) + 1):
    F[i][0] = ind * i

  for i in range(1, len(seqB) + 1):
    for j in range(1, len(seqA) + 1):
      s1 = F[i-1][j-1] + _subs(seqA[j-1], seqB[i-1], ide, sub)
      s2 = F[i-1][j] + ind
      s3 = F[i][j-1] + ind
      F[i][j] = max(s1, s2, s3)

  alignA = &#34;&#34;
  alignB = &#34;&#34;

  i = len(seqB)
  j = len(seqA)

  while(i &gt; 0 and j &gt; 0):
    score = F[i][j]
    scoreDiag = F[i-1][j-1]
    scoreUp = F[i-1][j]
    scoreLeft = F[i][j-1]
    if score == scoreDiag + _subs(seqA[j-1], seqB[i-1], ide, sub):
      alignA = seqA[j-1] + alignA
      alignB = seqB[i-1] + alignB
      i -= 1
      j -= 1
    elif score == scoreLeft + ind:
      alignA = seqA[j-1] + alignA
      alignB = &#34;-&#34; + alignB
      j -= 1
    elif score == scoreUp + ind:
      alignA = &#34;-&#34; + alignA
      alignB = seqB[i-1] + alignB
      i -= 1

  while(j &gt; 0):
    alignA = seqA[j-1] + alignA
    alignB = &#34;-&#34; + alignB
    j -= 1
  while(i &gt; 0):
    alignA = &#34;-&#34; + alignA
    alignB = seqB[i-1] + alignB
    i -= 1

  print(alignA)
  print(alignB)
  return (alignA, alignB)</code></pre>
</details>
</dd>
<dt id="miniDna.sequence.isDna"><code class="name flex">
<span>def <span class="ident">isDna</span></span>(<span>seq, msg=False)</span>
</code></dt>
<dd>
<section class="desc"><p>Test if a string is a DNA sequence.</p>
<p><strong>Keyword arguments:</strong><br>
seq &ndash; the string to test<br>
msg &ndash; display messages or not</p></section>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def isDna(seq: str, msg: bool = False) -&gt; bool: 
  &#34;&#34;&#34;Test if a string is a DNA sequence.

    **Keyword arguments:**  
    seq -- the string to test  
    msg -- display messages or not
  &#34;&#34;&#34;
  isRna = False
  for n in seq:
    if n not in NUCLEOTIDES:
      print(&#34;Invalid DNA sequence&#34;)
      return False
    elif n == &#39;U&#39; or n == &#39;u&#39;:
      isRna = True

  if isRna and msg:
    print(seq + &#34;\nis a RNA sequence&#34;)
  elif msg:
    print(seq + &#34;\nis a DNA sequence&#34;)
  return True</code></pre>
</details>
</dd>
<dt id="miniDna.sequence.percentIdentical"><code class="name flex">
<span>def <span class="ident">percentIdentical</span></span>(<span>seqA, seqB)</span>
</code></dt>
<dd>
<section class="desc"><p>Return the percentage of identity between two DNA sequences.</p>
<p><strong>Keyword arguments:</strong><br>
seqA &ndash; the first sequence
seqB &ndash; the second sequence</p></section>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def percentIdentical(seqA: str, seqB: str) -&gt; float:
  &#34;&#34;&#34;Return the percentage of identity between two DNA sequences.
    
    **Keyword arguments:**  
    seqA -- the first sequence
    seqB -- the second sequence
  &#34;&#34;&#34;

  if len(seqA) != len(seqB):
    raise ValueError(&#34;sequences have unequal length&#34;)
  dist = sum(ch1 != ch2 for ch1, ch2 in zip(seqA, seqB))
  return math.ceil(100 * (1 - dist/len(seqA)))</code></pre>
</details>
</dd>
<dt id="miniDna.sequence.simpleAlign"><code class="name flex">
<span>def <span class="ident">simpleAlign</span></span>(<span>seqA, seqB)</span>
</code></dt>
<dd>
<section class="desc"><p>Find the best way to superimpose two sequences
without changing them.
Empty nucleotides used to match as best as possible
the sequences are symbolized with a "-" character.</p>
<p><strong>Keyword arguments:</strong><br>
seqA &ndash; the first sequence<br>
seqB &ndash; the second sequence</p></section>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def simpleAlign(seqA: str, seqB: str) -&gt; tuple:
  &#34;&#34;&#34;Find the best way to superimpose two sequences
    without changing them.
    Empty nucleotides used to match as best as possible
    the sequences are symbolized with a &#34;-&#34; character.
    
    **Keyword arguments:**  
    seqA -- the first sequence  
    seqB -- the second sequence
  &#34;&#34;&#34;

  bestShift = 0
  shift = 0
  countId = countIdentical(seqA, seqB)

  while(shift &lt; len(seqA)):
    shift += 1
    a = seqA[shift:len(seqA)]
    b = seqB[0:len(seqB) - shift]
    c = countIdentical(a, b)
    if c &gt; countId:
      countId = c
      bestShift = shift

  
  finalA = seqA + &#34;&#34;.join(&#34;-&#34; for j in range(bestShift))
  finalB = &#34;&#34;.join(&#34;-&#34; for i in range(bestShift)) + seqB

  return (finalA, finalB)</code></pre>
</details>
</dd>
<dt id="miniDna.sequence.translate"><code class="name flex">
<span>def <span class="ident">translate</span></span>(<span>seq)</span>
</code></dt>
<dd>
<section class="desc"><p>find the protein coded in a DNA sequence</p>
<p><strong>Keyword arguments:</strong><br>
seq &ndash; DNA sequence to translate</p></section>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def translate(seq: str) -&gt; str:
  &#34;&#34;&#34;find the protein coded in a DNA sequence
    
    **Keyword arguments:**  
    seq -- DNA sequence to translate  
  &#34;&#34;&#34;

  start = 0
  protein = &#39;&#39;
  while start+2 &lt; len(seq):
    codon = seq[start:start+3]
    protein += AMINOCODE[codon]
    start += 3
  return protein</code></pre>
</details>
</dd>
</dl>
</section>
<section>
</section>
</article>
<nav id="sidebar">
<h1>Index</h1>
<div class="toc">
<ul></ul>
</div>
<ul id="index">
<li><h3>Super-module</h3>
<ul>
<li><code><a title="miniDna" href="index.html">miniDna</a></code></li>
</ul>
</li>
<li><h3><a href="#header-variables">Global variables</a></h3>
<ul class="">
<li><code><a title="miniDna.sequence.AMINO" href="#miniDna.sequence.AMINO">AMINO</a></code></li>
<li><code><a title="miniDna.sequence.AMINOCODE" href="#miniDna.sequence.AMINOCODE">AMINOCODE</a></code></li>
</ul>
</li>
<li><h3><a href="#header-functions">Functions</a></h3>
<ul class="two-column">
<li><code><a title="miniDna.sequence.compare" href="#miniDna.sequence.compare">compare</a></code></li>
<li><code><a title="miniDna.sequence.countIdentical" href="#miniDna.sequence.countIdentical">countIdentical</a></code></li>
<li><code><a title="miniDna.sequence.display" href="#miniDna.sequence.display">display</a></code></li>
<li><code><a title="miniDna.sequence.dotPlot" href="#miniDna.sequence.dotPlot">dotPlot</a></code></li>
<li><code><a title="miniDna.sequence.filterDotPlot" href="#miniDna.sequence.filterDotPlot">filterDotPlot</a></code></li>
<li><code><a title="miniDna.sequence.freqAt" href="#miniDna.sequence.freqAt">freqAt</a></code></li>
<li><code><a title="miniDna.sequence.freqList" href="#miniDna.sequence.freqList">freqList</a></code></li>
<li><code><a title="miniDna.sequence.globalAlign" href="#miniDna.sequence.globalAlign">globalAlign</a></code></li>
<li><code><a title="miniDna.sequence.isDna" href="#miniDna.sequence.isDna">isDna</a></code></li>
<li><code><a title="miniDna.sequence.percentIdentical" href="#miniDna.sequence.percentIdentical">percentIdentical</a></code></li>
<li><code><a title="miniDna.sequence.simpleAlign" href="#miniDna.sequence.simpleAlign">simpleAlign</a></code></li>
<li><code><a title="miniDna.sequence.translate" href="#miniDna.sequence.translate">translate</a></code></li>
</ul>
</li>
</ul>
</nav>
</main>
<footer id="footer">
<p>Generated by <a href="https://pdoc3.github.io/pdoc"><cite>pdoc</cite> 0.7.1</a>.</p>
</footer>
<script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/highlight.min.js"></script>
<script>hljs.initHighlightingOnLoad()</script>
</body>
</html>